source/DamageProfile.cpp

```diff
@@ -0,0 +1,148 @@
+/* DamageProfile.cpp
+Copyright (c) 2022 by Amazinite
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/
+
+#include "DamageProfile.h"
+
+#include "Mask.h"
+#include "Outfit.h"
+#include "Ship.h"
+#include "Weapon.h"
+
+using namespace std;
+
+DamageProfile::DamageProfile(const Projectile::ImpactInfo &info, double damageScaling, bool isBlast, bool skipFalloff)
+	: weapon(info.weapon), position(info.position), distanceTraveled(info.distanceTraveled),
+	inputScaling(damageScaling), isBlast(isBlast), skipFalloff(skipFalloff)
+{
+	// Precompute blast damage scaling.
+	if(isBlast && weapon.IsDamageScaled())
+	{
+		// Scale blast damage based on the distance from the blast
+		// origin and if the projectile uses a trigger radius. The
+		// point of contact must be measured on the sprite outline.
+		// scale = (1 + (tr / (2 * br))^2) / (1 + r^4)^2
+		double blastRadius = max(1., weapon.BlastRadius());
+		double radiusRatio = weapon.TriggerRadius() / blastRadius;
+		k = !radiusRatio ? 1. : (1. + .25 * radiusRatio * radiusRatio);
+		rSquared = 1. / (blastRadius * blastRadius);
+	}
+	if(!skipFalloff && weapon.HasDamageDropoff())
+		inputScaling *= weapon.DamageDropoff(distanceTraveled);
+}
+
+
+
+// Set a distance traveled to be used on the next CalculateDamage call,
+// assuming skipFalloff is true.
+void DamageProfile::SetDistance(double distance)
```

Comment:
Could we move the distance scaling to even earlier in the process (as suggested in #6562). I feel that it makes the code even simpler.

Reply:
Not sure what you mean by that. For damage taken from projectiles, the distance scaling is already only applied once for all impacted ships within the DamageProfile constructor, while for damage taken from hazards it's done for each ship. If you mean moving the distance scaling outside of DamageProfile, I don't see why we'd do that when the whole point of DamageProfile is to be the object that does the damage calculations.